### PR TITLE
Fix REL32 relocs in COFF ##bin

### PR DESCRIPTION
**Pull request by pelijah**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

File: libr/bin/p/bin_coff.c [hunk 1]
```diff
@@ -345,7 +345,10 @@ static RList *_relocs_list(RBin *rbin, struct r_bin_coff_obj *bin, bool patch, u
 					case COFF_REL_I386_REL32:
 						reloc->type = R_BIN_RELOC_32;
 						reloc->additive = 1;
-						ut64 data = sym_vaddr - reloc->vaddr - 4;
+						ut8 offset[4];
+						rbin->iob.read_at (rbin->iob.io, reloc->vaddr, offset, 4);
```
Comment (ret2libc) [inline libr/bin/p/bin_coff.c:349]: check for return value and print a warning and/or assign a default value to `data`?

File: libr/bin/p/bin_coff.c [hunk 2]
```diff
@@ -345,7 +345,12 @@ static RList *_relocs_list(RBin *rbin, struct r_bin_coff_obj *bin, bool patch, u
 					case COFF_REL_I386_REL32:
 						reloc->type = R_BIN_RELOC_32;
 						reloc->additive = 1;
-						ut64 data = sym_vaddr - reloc->vaddr - 4;
+						ut8 offset[4];
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:348]: just = {0}; instead of the memset below

File: libr/bin/p/bin_coff.c [hunk 3]
```diff
@@ -345,7 +345,12 @@ static RList *_relocs_list(RBin *rbin, struct r_bin_coff_obj *bin, bool patch, u
...
+						if (!rbin->iob.read_at (rbin->iob.io, reloc->vaddr, offset, 4)) {
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:349]: if the read fails you shuoldnt move forward imho

File: libr/bin/p/bin_coff.c [hunk 4]
```diff
@@ -356,7 +361,12 @@ static RList *_relocs_list(RBin *rbin, struct r_bin_coff_obj *bin, bool patch, u
 					case COFF_REL_AMD64_REL32:
 						reloc->type = R_BIN_RELOC_32;
 						reloc->additive = 1;
-						ut64 data = sym_vaddr - reloc->vaddr - 4;
+						ut8 offset[4];
+						if (!rbin->iob.read_at (rbin->iob.io, reloc->vaddr, offset, 4)) {
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:365]: same here, i think you can just avoid doing the rest of lines in this case if the read cant happen

File: test/db/formats/coff [hunk 5]
```diff
@@ -133,3 +133,16 @@ sym._TIFFGetField 0xb469 [CALL] call sym._TIFFVGetField
...
 RUN
+
+NAME=patching REL32 amd64
+ARGS=-e io.cache=1
+FILE=bins/coff/coff.obj
+CMDS=<<EOF
+pd 2 @ sym.main+6
+EOF
+EXPECT=<<EOF
+            0x0000006a      4c8d051a0000.  lea r8, [0x0000008b]        ; sym..data ; "Win64 assembly"; RELOC 32 .data @ 0x0000008b + 0x0
```
Comment (trufae) [inline test/db/formats/coff:144]: this + 0x0 makes no sense. can we get rid of them?

Comment (pelijah) [reply test/db/formats/coff:144]: Of course but it requires changes in `core/disasm.c` and isn't related to the current PR.

Comment (trufae) [reply test/db/formats/coff:144]: Good, then looks good to merge, can you open a separate PR for this? Thanks!

ill fix the PR title now

Comment (pelijah) [reply test/db/formats/coff:144]: #17312

**Pull request merged**